File: connectors/openclaw/src/openclaw/media.ts
/**
 * #158 出站附件 + #372 capability-bound 路徑校驗:
 * agent 回覆正文裡 `MEDIA:<路徑>` 標記 → 讀文件(12MB 上限)→ media.attach 事件
 * (base64 內聯,server 落存儲+渲染 media block)。
 *
 * 威脅:模型正文不是可信授權邊界;prompt injection 可誘導打印 `/etc/hosts`、
 * `~/.ssh/*` 等,舊實現只要文件存在就上傳。
 *
 * 對策(#372 務實落地):
 *  - **只認 `MEDIA:` 標記**(禁用裸絕對路徑自動上傳,降低誤報與注入面);
 *  - 路徑必須落在「會話允許根」內:session cwd / macchiato 管理目錄 / 附件目錄 /
 *    `MACCHIATO_MEDIA_ALLOW_ROOTS`(冒號分隔額外根,測試與本機擴展用);
 *  - `realpath` 解析 symlink 後做包含性檢查(逃逸根外 → 拒);
 *  - `O_NOFOLLOW` 打開 + `fstat` 確認常規文件 + size 上限(縮 TOCTOU)。
 * 失敗只記 reason / path_len,不 dump 路徑正文或文件內容(#381 對齊)。
 */
import {
  openSync,
  readSync,
  fstatSync,
  closeSync,
  lstatSync,
  realpathSync,
  statSync,
  constants,
} from "node:fs";
import { basename, extname, join, sep } from "node:path";
import { homedir } from "node:os";

export const MEDIA_MAX = 12 * 1024 * 1024;
/** 冒號分隔的額外允許根(測試 / 本機 agent 輸出目錄白名單)。 */
export const MEDIA_ALLOW_ROOTS_ENV = "MACCHIATO_MEDIA_ALLOW_ROOTS";

const MIME: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".svg": "image/svg+xml",
  ".pdf": "application/pdf",
  ".txt": "text/plain",
  ".md": "text/markdown",
  ".csv": "text/csv",
  ".json": "application/json",
  ".html": "text/html",
  ".zip": "application/zip",
  ".mp3": "audio/mpeg",
  ".m4a": "audio/mp4",
  ".wav": "audio/wav",
  ".mp4": "video/mp4",
  ".mov": "video/quicktime",
};

/** MEDIA: 標記(整行,顯式意圖)。裸絕對路徑不再自動上傳(#372)。 */
const MEDIA_RE = /^MEDIA:\s*(\S.*?)\s*$/gm;

const OPEN_FLAGS = constants.O_RDONLY | (constants.O_NOFOLLOW ?? 0);

function expand(p: string): string {
  if (p === "~") return homedir();
  if (p.startsWith("~/")) return join(homedir(), p.slice(2));
  return p;
}

/** target 是否在 root 之內(邊界安全:同路徑或以 root+sep 為前綴,防 /a/bc 誤判在 /a/b 內)。 */
export function withinRoot(root: string, target: string): boolean {
  if (target === root) return true;
  const prefix = root.endsWith(sep) ? root : root + sep;
  return target.startsWith(prefix);
}

/**
 * 彙總當前會話的允許根(realpath 後的目錄;不存在/不可解析的跳過)。
 * 空列表 = fail closed(任何路徑都拒)。
 */
export function resolveMediaAllowRoots(opts?: {
  sessionCwd?: string | null;
  extraRoots?: readonly string[];
}): string[] {
  const candidates: string[] = [];
  const cwd = opts?.sessionCwd?.trim();
  if (cwd) candidates.push(expand(cwd));

  const state =
    (process.env.MACCHIATO_STATE_DIR || "").trim() || join(homedir(), ".macchiato");
  candidates.push(state);
  candidates.push(join(state, "attachments"));

  const raw = (process.env[MEDIA_ALLOW_ROOTS_ENV] || "").trim();
  if (raw) {
    for (const part of raw.split(":")) {
      const p = part.trim();
      if (p) candidates.push(expand(p));
    }
  }
  if (opts?.extraRoots) {
    for (const r of opts.extraRoots) {
      const p = r?.trim();
      if (p) candidates.push(expand(p));
    }
  }

  const out: string[] = [];
  const seen = new Set<string>();
  for (const c of candidates) {
    try {
      const canon = realpathSync(c);
      if (!statSync(canon).isDirectory()) continue;
      if (seen.has(canon)) continue;
      seen.add(canon);
      out.push(canon);
    } catch {
      /* 不存在 / 不可解析 → 跳過該根 */
    }
  }
  return out;
}

/**
 * 將候選路徑規範化並校驗是否落在允許根內且為常規文件。
 * 成功 → realpath;失敗 → null(不拋,由調用方記 reason)。
 */
export function resolveAllowedMediaPath(
  rawPath: string,
  roots: readonly string[],
): string | null {
  if (!roots.length) return null;
  const expanded = expand(rawPath.trim());
  if (!expanded) return null;
  let canon: string;
  try {
    canon = realpathSync(expanded);
  } catch {
    return null;
  }
  if (!roots.some((r) => withinRoot(r, canon))) return null;
  try {
    // realpath 之後末段必非 symlink;再 lstat 確認是常規文件(拒目錄/裝置/socket)。
    const ls = lstatSync(canon);
    if (!ls.isFile()) return null;
  } catch {
    return null;
  }
  return canon;
}

function logDenied(reason: string, pathLen: number): void {
  console.error(`[media denied] reason=${reason} path_len=${pathLen}`);
}

/** 從回覆正文提取待投遞的本地文件路徑(僅 MEDIA: ;去重;必須落在允許根內)。 */
export function extractMediaPaths(text: string, roots: readonly string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  if (!roots.length) return out;
  for (const m of text.matchAll(MEDIA_RE)) {
    const raw = (m[1] ?? "").trim();
    if (!raw) continue;
    const canon = resolveAllowedMediaPath(raw, roots);
    if (!canon || seen.has(canon)) continue;
    seen.add(canon);
    out.push(canon);
  }
  return out;
}

export interface MediaPayload {
  kind: "image" | "document";
  name: string;
  mime: string;
  size: number;
  data_b64: string;
}

/** 讀文件成 media.attach payload;根外/超限/非常規/讀失敗 → null。 */
export function readMediaFile(path: string, roots: readonly string[]): MediaPayload | null {
  const pathLen = path.length;
  const canon = resolveAllowedMediaPath(path, roots);
  if (!canon) {
    logDenied("outside_roots_or_missing", pathLen);
    return null;
  }

  let fd: number;
  try {
    fd = openSync(canon, OPEN_FLAGS);
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === "ELOOP") logDenied("symlink_race", pathLen);
    else logDenied("open_failed", pathLen);
    return null;
  }

  try {
    const st = fstatSync(fd);
    if (!st.isFile()) {
      logDenied("not_regular_file", pathLen);
      return null;
    }
    const size = Number(st.size);
    if (size <= 0) {
      logDenied("empty", pathLen);
      return null;
    }
    if (size > MEDIA_MAX) {
      console.error(`[media too big] size=${size}B path_len=${pathLen}`);
      return null;
    }
    const buf = Buffer.allocUnsafe(size);
    let off = 0;
    while (off < size) {
      const n = readSync(fd, buf, off, size - off, off);
      if (n <= 0) break;
      off += n;
    }
    if (off !== size) {
      logDenied("short_read", pathLen);
      return null;
    }
    const mime = MIME[extname(canon).toLowerCase()] ?? "application/octet-stream";
    return {
      kind: mime.startsWith("image/") ? "image" : "document",
      name: basename(canon),
      mime,
      size,
      data_b64: buf.toString("base64"),
    };
  } catch {
    logDenied("read_failed", pathLen);
    return null;
  } finally {
    closeSync(fd);
  }
}

File: connectors/hermes/media.py
"""#372 出站附件路徑 capability-bound 校驗。

背景/威脅:agent 正文裡 `MEDIA:<path>` 舊實現只要文件存在就讀並 media.attach。
模型輸出不是可信授權邊界;prompt injection 可誘導打印 `/etc/hosts`、`~/.ssh/*` 等。

對策(務實落地,對齊 openclaw media.ts):
  - **只認 `MEDIA:` 標記**(不再合併裸絕對路徑自動上傳);
  - 路徑必須落在「會話允許根」內:
      · 會話 cwd(drive 有 per-session cwd 時);
      · connector 管理目錄(STATE_DIR / ATTACH_DIR);
      · `MACCHIATO_MEDIA_ALLOW_ROOTS`(冒號分隔額外根,測試與本機擴展);
  - realpath 解析 symlink 後做包含性檢查(逃逸根外 → 拒);
  - O_NOFOLLOW 打開 + fstat 常規文件 + size 上限(縮 TOCTOU)。
失敗只記 reason / path_len,不 dump 路徑正文或文件內容。
"""

from __future__ import annotations

import base64
import mimetypes
import os
import stat
import sys
from typing import Optional

MEDIA_MAX = 12 * 1024 * 1024  # 出站附件 base64 內聯上限(12MB),超出跳過
MEDIA_ALLOW_ROOTS_ENV = "MACCHIATO_MEDIA_ALLOW_ROOTS"


def expand_home(p: str) -> str:
    """~ / ~/... 展開為絕對路徑(不觸碰 fs 解析 symlink)。"""
    if p == "~":
        return os.path.expanduser("~")
    if p.startswith("~/"):
        return os.path.join(os.path.expanduser("~"), p[2:])
    return p


def within_root(root: str, target: str) -> bool:
    """target 是否在 root 之內(邊界安全:同路徑或以 root+sep 為前綴)。"""
    if target == root:
        return True
    prefix = root if root.endswith(os.sep) else root + os.sep
    return target.startswith(prefix)


def resolve_media_allow_roots(
    *,
    session_cwd: Optional[str] = None,
    state_dir: Optional[str] = None,
    attach_dir: Optional[str] = None,
    extra_roots: Optional[list[str]] = None,
) -> list[str]:
    """彙總允許根(realpath 後的目錄;不存在/不可解析跳過)。空列表 = fail closed。"""
    candidates: list[str] = []
    cwd = (session_cwd or "").strip()
    if cwd:
        candidates.append(expand_home(cwd))

    # connector 管理目錄(調用方可注入;默認從 env / ~/.macchiato 推)
    if state_dir is None:
        state_dir = (os.environ.get("MACCHIATO_STATE_DIR") or "").strip() or "~/.macchiato"
    state_abs = expand_home(state_dir)
    candidates.append(state_abs)

    if attach_dir is None:
        attach_abs = os.path.join(state_abs, "attachments")
    else:
        attach_abs = expand_home(attach_dir)
    candidates.append(attach_abs)

    raw = (os.environ.get(MEDIA_ALLOW_ROOTS_ENV) or "").strip()
    if raw:
        for part in raw.split(":"):
            p = part.strip()
            if p:
                candidates.append(expand_home(p))
    if extra_roots:
        for r in extra_roots:
            p = (r or "").strip()
            if p:
                candidates.append(expand_home(p))

    out: list[str] = []
    seen: set[str] = set()
    for c in candidates:
        try:
            if not os.path.exists(c):
                continue
            canon = os.path.realpath(c)
            if not os.path.isdir(canon):
                continue
            if canon in seen:
                continue
            seen.add(canon)
            out.append(canon)
        except OSError:
            pass
    return out


def resolve_allowed_media_path(raw_path: str, roots: list[str]) -> Optional[str]:
    """規範化並校驗:realpath + 落在某根內 + 常規文件。成功回 realpath,否則 None。"""
    if not roots:
        return None
    expanded = expand_home((raw_path or "").strip())
    if not expanded:
        return None
    try:
        canon = os.path.realpath(expanded)
    except OSError:
        return None
    if not any(within_root(r, canon) for r in roots):
        return None
    try:
        st = os.lstat(canon)  # realpath 後末段必非 symlink
        if not stat.S_ISREG(st.st_mode):
            return None
    except OSError:
        return None
    return canon


def _log_denied(reason: str, path_len: int) -> None:
    print(f"[media denied] reason={reason} path_len={path_len}", file=sys.stderr)


def extract_media_files(text: str, allowed_roots: list[str]) -> list[str]:
    """從正文提取 MEDIA: 路徑(去重、存在性+允許根校驗)。不認裸絕對路徑。

    復用 Hermes 平台適配器的 MEDIA: 解析(與 Discord/Telegram 投遞同一套標記語法),
    但**不**合併 extract_local_files 的裸路徑(#372)。
    """
    if not allowed_roots:
        return []
    try:
        from gateway.platforms.base import BasePlatformAdapter as B
    except ImportError:
        # 無 Hermes 時僅做極簡 MEDIA: 行解析(單測 / 無 gateway 環境)
        return _extract_media_markers_fallback(text, allowed_roots)

    media, _cleaned = B.extract_media(text)
    media = B.filter_media_delivery_paths(media)  # [(path, is_voice)],校驗存在
    out: list[str] = []
    seen: set[str] = set()
    for path, _v in media:
        canon = resolve_allowed_media_path(path, allowed_roots)
        if not canon or canon in seen:
            continue
        seen.add(canon)
        out.append(canon)
    return out


def _extract_media_markers_fallback(text: str, allowed_roots: list[str]) -> list[str]:
    """無 Hermes 時的 MEDIA: 標記解析(行首或內聯 MEDIA:path)。"""
    import re

    out: list[str] = []
    seen: set[str] = set()
    # 行首整行 或 內聯 MEDIA:/abs/path(與歷史 Hermes 測試 `see MEDIA:{p} done` 對齊)
    for m in re.finditer(r"(?:^|\s)MEDIA:\s*(\S+)", text or "", flags=re.M):
        canon = resolve_allowed_media_path(m.group(1), allowed_roots)
        if not canon or canon in seen:
            continue
        seen.add(canon)
        out.append(canon)
    return out


def read_media_file(path: str, allowed_roots: list[str]) -> Optional[dict]:
    """讀文件成 media.attach payload;根外/超限/非常規/讀失敗 → None。"""
    path_len = len(path or "")
    canon = resolve_allowed_media_path(path, allowed_roots)
    if not canon:
        _log_denied("outside_roots_or_missing", path_len)
        return None

    flags = os.O_RDONLY
    if hasattr(os, "O_NOFOLLOW"):
        flags |= os.O_NOFOLLOW
    try:
        fd = os.open(canon, flags)
    except OSError as exc:
        reason = "symlink_race" if getattr(exc, "errno", None) in (
            getattr(__import__("errno"), "ELOOP", -1),
        ) else "open_failed"
        _log_denied(reason, path_len)
        return None

    try:
        st = os.fstat(fd)
        if not stat.S_ISREG(st.st_mode):
            _log_denied("not_regular_file", path_len)
            return None
        size = int(st.st_size)
        if size <= 0:
            _log_denied("empty", path_len)
            return None
        if size > MEDIA_MAX:
            print(f"[media too big] size={size}B path_len={path_len}", file=sys.stderr)
            return None
        chunks: list[bytes] = []
        remaining = size
        while remaining > 0:
            buf = os.read(fd, remaining)
            if not buf:
                break
            chunks.append(buf)
            remaining -= len(buf)
        data = b"".join(chunks)
        if len(data) != size:
            _log_denied("short_read", path_len)
            return None
        mime = mimetypes.guess_type(canon)[0] or "application/octet-stream"
        return {
            "kind": "image" if mime.startswith("image/") else "document",
            "name": os.path.basename(canon),
            "mime": mime,
            "size": size,
            "data_b64": base64.b64encode(data).decode("ascii"),
        }
    except OSError:
        _log_denied("read_failed", path_len)
        return None
    finally:
        try:
            os.close(fd)
        except OSError:
            pass
